presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxByNAggregation.java

```diff
@@ -0,0 +1,177 @@
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.operator.aggregation.AggregationUtils.generateAggregationName;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractMinMaxByNAggregation
+        extends ParametricAggregation
+{
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "input", BlockComparator.class, Type.class, Type.class, MinMaxByNState.class, Block.class, Block.class, int.class, long.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "combine", MinMaxByNState.class, MinMaxByNState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(AbstractMinMaxByNAggregation.class, "output", ArrayType.class, MinMaxByNState.class, BlockBuilder.class);
+
+    private final String name;
+    private final Function<Type, BlockComparator> typeToComparator;
+    private final Signature signature;
+
+    protected AbstractMinMaxByNAggregation(String name, Function<Type, BlockComparator> typeToComparator)
+    {
+        this.name = requireNonNull(name);
+        this.typeToComparator = requireNonNull(typeToComparator);
+        this.signature = new Signature(name, ImmutableList.of(typeParameter("V"), orderableTypeParameter("K")), "array<V>", ImmutableList.of("V", "K", StandardTypes.BIGINT), false, false);
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return signature;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type keyType = types.get("K");
+        Type valueType = types.get("V");
+        Signature signature = new Signature(name, new ArrayType(valueType).getTypeSignature(), valueType.getTypeSignature(), keyType.getTypeSignature(), BIGINT.getTypeSignature());
+        InternalAggregationFunction aggregation = generateAggregation(valueType, keyType);
+        return new FunctionInfo(signature, getDescription(), aggregation);
+    }
+
+    public static void input(BlockComparator comparator, Type valueType, Type keyType, MinMaxByNState state, Block value, Block key, int blockIndex, long n)
+    {
+        TypedKeyValueHeap heap = state.getTypedKeyValueHeap();
+        if (heap == null) {
+            if (n <= 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "second argument of max_by_n/min_by_n must be positive");
```

Comment:
the functions are named `max_by` and `min_by`, so remove the `_n` from this string